### PR TITLE
Avijit/bazel static support

### PR DESCRIPTION
**Pull request by kanvi-nervana**
Add support for building the c++ application using bazel and linking statically as opposed to dynamic linking.

### CODE REVIEW

**Comment by sindhu-nervana** [comment]
@googlebot I fixed it

**Comment by sindhu-nervana** [comment]
@googlebot I fixed it

**Comment by kanvi-nervana** [comment]
@googlebot I fixed it.

**Comment by sindhu-nervana** [comment]
@googlebot I fixed it.

**Comment by kanvi-nervana** [comment]
> This has commits from `sindhura.kantamani` and `sindhu-nervana`.
> `sindhu-nervana` can pass CLA but `sindhura.kantamani` can't I think. So we may have to resubmit the whole PR to make it pass CLA?

@sayantan-nervana  Yes, we tried fixing it. We'll have to create a new PR.

File: bazel/WORKSPACE [hunk 1]
```diff
@@ -15,13 +15,43 @@
...
-load("@//tf_configure:tf_configure.bzl", "tf_configure")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
+load("@//tf_configure:tf_configure.bzl", "tf_configure")
 tf_configure(
     name = "local_config_tf",
 )
 
+# TensorFlow depends on "io_bazel_rules_closure" so we need this here.
+# Needs to be kept in sync with the same target in TensorFlow's WORKSPACE file.
```
Comment (sayantan-nervana) [inline bazel/WORKSPACE]: We should document this requirement here (TF change/upgrade checklist): https://github.com/tensorflow/ngraph-bridge/wiki/Update-TF-version-in-ngraph-bridge

File: bazel/ngraph.BUILD [hunk 2]
```diff
@@ -168,7 +119,7 @@ cc_library(
         "-fstack-protector-all",
         '-D SHARED_LIB_PREFIX=\\"lib\\"',
         '-D SHARED_LIB_SUFFIX=\\".so\\"',
-        '-D NGRAPH_VERSION=\\"v0.20.0-rc.0\\"',
+        '-D NGRAPH_VERSION=\\"v0.25.1-rc.0\\"',
```
Comment (sayantan-nervana) [inline bazel/ngraph.BUILD]: We should document that these ngraph versions need to change in this checklist: https://github.com/tensorflow/ngraph-bridge/wiki/Release-Checklist

File: bazel/ngraph.BUILD [hunk 3]
```diff
@@ -82,73 +85,21 @@ cc_library(
...
```
Comment (sayantan-nervana) [inline bazel/ngraph.BUILD]: We should document that these ngraph versions need to change in this checklist: https://github.com/tensorflow/ngraph-bridge/wiki/Release-Checklist

File: bazel/ngraph.BUILD [hunk 4]
```diff
@@ -314,17 +270,20 @@ cc_binary(
         "-fstack-protector-all",
         '-D SHARED_LIB_PREFIX=\\"lib\\"',
         '-D SHARED_LIB_SUFFIX=\\".so\\"',
-        '-D NGRAPH_VERSION=\\"0.19.0-rc.4\\"',
+        '-D NGRAPH_VERSION=\\"0.25.1-rc.0\\"',
```
Comment (sayantan-nervana) [inline bazel/ngraph.BUILD]: We should document that these ngraph versions need to change in this checklist: https://github.com/tensorflow/ngraph-bridge/wiki/Release-Checklist

File: examples/multisession/sample.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  tensorflow::SessionOptions options;
+  options.config.mutable_graph_options()
+      ->mutable_optimizer_options()
+      ->set_opt_level(tensorflow::OptimizerOptions_Level_L0);
+  options.config.mutable_graph_options()
+      ->mutable_rewrite_options()
+      ->set_constant_folding(tensorflow::RewriterConfig::OFF);
+
+#if 0
+  // The following is related to Grapller - which we are turning off
```
Comment (sayantan-nervana) [inline examples/multisession/sample.cpp]:
```suggestion
  // The following is related to Grappler - which we are turning off
```

File: examples/multisession/sample.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  // Turn off optimizations so that all the nodes are processed
+  tensorflow::SessionOptions options;
+  options.config.mutable_graph_options()
+      ->mutable_optimizer_options()
+      ->set_opt_level(tensorflow::OptimizerOptions_Level_L0);
+  options.config.mutable_graph_options()
+      ->mutable_rewrite_options()
+      ->set_constant_folding(tensorflow::RewriterConfig::OFF);
+
+#if 0
```
Comment (sayantan-nervana) [inline examples/multisession/sample.cpp]: Should we use `NGRAPH_TF_USE_GRAPPLER_OPTIMIZER` here?

File: examples/multisession/sample.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  tensorflow::ClientSession session(root, options);
+
+  std::vector<tensorflow::Tensor> outputs;
+  session.Run({R}, &outputs);
+
+  // Print the output
+  std::cout << "Result: " << outputs[0].matrix<float>() << std::endl;
+
+  // Close the session.
+  // session.Close();
```
Comment (sayantan-nervana) [inline examples/multisession/sample.cpp]: the close line is commented out

File: examples/cpp/hello_tf.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    cout << "Backend: " << backend_name << std::endl;
+  }
+}
+
+// Sets the specified backend. This backend must be set BEFORE running
+// the computation
+tensorflow::Status SetNGraphBackend(const string& backend_name) {
+  // Select a backend
+  tensorflow::Status status =
+      tensorflow::ngraph_bridge::BackendManager::SetBackendName(backend_name);
```
Comment (sayantan-nervana) [inline examples/cpp/hello_tf.cpp]: This style of backend setting does not work in grappler style builds I think. For grappler builds, we have to use the rewriter config

File: bazel/BUILD [hunk 9]
```diff
@@ -114,3 +158,86 @@ cc_binary(
     ],
     visibility = ["//visibility:public"],
 )
+
+load("@org_tensorflow//tensorflow:tensorflow.bzl", "tf_cc_binary")
+tf_cc_binary(
```
Comment (sayantan-nervana) [inline bazel/BUILD]: Should we have an UT, where we `bazel build hello_tf` and then run the executable?
That keeps the "bazel building of executables" path under test in CI

File: bazel/ngraph.BUILD (hunk 2 above)
Comment (kanvi-nervana) [reply bazel/ngraph.BUILD]: Already did.

File: bazel/ngraph.BUILD (hunk 3 above)
Comment (kanvi-nervana) [reply bazel/ngraph.BUILD]: Already did.

File: bazel/ngraph.BUILD (hunk 4 above)
Comment (kanvi-nervana) [reply bazel/ngraph.BUILD]: Already did.

File: bazel/WORKSPACE (hunk 1 above)
Comment (kanvi-nervana) [reply bazel/WORKSPACE]: Done

**Pull request closed without merging**